src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceCollectionServiceExtensionsTest.cs

```diff
@@ -414,5 +414,91 @@ public void RemoveAll_RemovesAllServicesWithMatchingServiceType()
             // Assert
             Assert.Equal(new[] { descriptor }, collection);
         }
+
+        [Fact]
+        public void IndexerWorks()
+        {
+            // Arrange
+            var descriptor = new ServiceDescriptor(typeof(IFakeServiceInstance), typeof(FakeService), ServiceLifetime.Transient);
+            var collection = new ServiceCollection
+            {
+                new ServiceDescriptor(typeof(IFakeService), typeof(FakeService), ServiceLifetime.Transient),
+                new ServiceDescriptor(typeof(IFakeService), typeof(FakeService), ServiceLifetime.Transient)
+            };
+
+            var existing = collection[1];
+
+            // Act
+            collection[1] = descriptor;
+
+            // Assert
+            Assert.False(collection.Contains(existing));
+            Assert.Same(collection[1], descriptor);
+        }
+
+        [Fact]
+        public void RemoveAtWorks()
+        {
+            // Arrange
+            var descriptor = new ServiceDescriptor(typeof(IFakeServiceInstance), typeof(FakeService), ServiceLifetime.Transient);
+            var collection = new ServiceCollection
+            {
+                new ServiceDescriptor(typeof(IFakeService), typeof(FakeService), ServiceLifetime.Transient),
+                descriptor,
+            };
+
+            var existing = collection[1];
+
+            // Act
+            collection.RemoveAt(1);
+
+            // Assert
+            Assert.False(collection.Contains(existing));
+            Assert.Equal(1, collection.Count);
+        }
+
+        [Fact]
+        public void InsertAtWorks()
+        {
+            // Arrange
+            var descriptor = new ServiceDescriptor(typeof(IFakeServiceInstance), typeof(FakeService), ServiceLifetime.Transient);
+            var collection = new ServiceCollection
+            {
+                new ServiceDescriptor(typeof(IFakeService), typeof(FakeService), ServiceLifetime.Transient)
+            };
+
+            // Act
+            collection.Insert(0, descriptor);
+
+            // Assert
+            Assert.Same(descriptor, collection[0]);
+            Assert.Equal(0, collection.IndexOf(descriptor));
+        }
+
+        [Fact]
+        public void InsertAtFollowedByReplaceWorks()
+        {
+            // Arrange
+            var aDescriptor = new ServiceDescriptor(typeof(IFakeService), typeof(A), ServiceLifetime.Transient);
+            var bDescriptor = new ServiceDescriptor(typeof(IFakeService), typeof(B), ServiceLifetime.Transient);
+            var cDescriptor = new ServiceDescriptor(typeof(IFakeService), typeof(C), ServiceLifetime.Transient);
+
+            var collection = new ServiceCollection
+            {
+                bDescriptor
+            };
+
+            // Act
+            collection.Insert(0, aDescriptor);
+            collection.Replace(cDescriptor);
+
+            // Assert
+            Assert.Equal(new[] { bDescriptor, cDescriptor }, collection);
```

Comment:
This test shows how weird the Replace API works.

1. Start with: `[b]`
2. Insert `a`: `[a, b]`
3. Replace with `c`: `[b, c]`

`a` got replaced, but `c` got added to the end. And `b` is still there.

Reply:
Yea, I couldn't figure out why replace was so weird until I was re-implementing it.